### ISSUE TITLE
ETAG/If-None-Match expalanation

### ISSUE THREAD

**Issue by AAverin** [square/okhttp#1278]
This is a copy of the question from Retrofit thread.
Also have a question on SO: stackoverflow.com/questions/27561992/retrofit-etag-and-caching

Is there a proper explanation on how to add caching and ETAG/If-None-Match support to Retrofit+OkHttp?
I'm struggling adding Etag support on 2 projects, and if at first I suspected that there might be an issue with HTTP Headers, another project has everything set correctly and caching still doesn't work as expected.

Following are my attempts to make it work. Results show that caching seems to be working within the same instance of the application, but as soon as I restart - everything loads long again.
Also, in my logs I didn't see If-None-Match being added to a request, so I assume that server isn't aware of ETag and still recalculates the response completely.

Here are some code samples:

     public class RetrofitHttpClient extends UrlConnectionClient {

        private OkUrlFactory generateDefaultOkUrlFactory() {
            OkHttpClient client = new com.squareup.okhttp.OkHttpClient();

            try {
                Cache responseCache = new Cache(baseContext.getCacheDir(), SIZE_OF_CACHE);
                client.setCache(responseCache);
            } catch (Exception e) {
                Logger.log(this, e, "Unable to set http cache");
            }

            client.setConnectTimeout(READ_TIMEOUT, TimeUnit.MILLISECONDS);
            client.setReadTimeout(CONNECT_TIMEOUT, TimeUnit.MILLISECONDS);
            return new OkUrlFactory(client);
        }

        private final OkUrlFactory factory;

        public RetrofitHttpClient() {
            factory = generateDefaultOkUrlFactory();
        }

        @Override
        protected HttpURLConnection openConnection(retrofit.client.Request request) throws IOException {
            return factory.open(new URL(request.getUrl()));
        }
    }

Rest adapter is then created with FULL log level and a custom tag:

    restAdapter = new RestAdapter.Builder()
                .setClient(new RetrofitHttpClient())
                .setEndpoint(Config.BASE_URL)
                .setRequestInterceptor(new SignatureSetter())
                .setConverter(new JacksonConverter(JsonHelper.getObjectMapper()))
                .setLogLevel(RestAdapter.LogLevel.FULL)
                .setLog(new AndroidLog("=NETWORK="))
                .build();

I have a long request on the first screen of the app for testing.
When I open the app - it takes 7 seconds to complete the request. If I pause and resume the app - same request takes 250ms, clearly hitting the cache. If I close the app completely and restart - it again takes 7 seconds.

In the end, when I would like to see If-None-Match in the logged headers so I would know for sure that Etag that was in the response is correctly used by the cache. Or a definitive answer if I need to implement etag caching logic manually.

**Comment by crossle** [comment]
I'm not see ETag header on Full level log, but curl print ETag

**Comment by swankjesse** [comment]
This is a question for Stack Overflow. Our issue tracker is for reporting bugs & requesting feature requests. Bug reports should come with test cases.

**Comment by AAverin** [comment]
As I mentioned, here's the SO question: http://stackoverflow.com/questions/27561992/retrofit-etag-and-caching
I was said to ask this question here in Retrofit thread - Jake said question is okhttp related, not retrofit

**Comment by jam-le** [comment]
I am having issues here as well. Pretty sure this is broken. If not, documentation surrounding usage would be super helpful, but it appears to be lacking.

**Comment by jam-le** [comment]
I did find a possible work-around. Some blog posts state that the etag matching should be automatic if set up properly. I found that, along with the etag field in the response, there is an "x-cache" field which will return "HIT" or "MISS". I have come to the conclusion that a cache hit indicates that a matching etag was found and a cache miss indicates that a matching etag was not found. This is what I have concluded from observing the responses returned under different scenarios.
